File: Shaders/Shader.lua
local ____lualib = require("lualib_bundle")
local __TS__Class = ____lualib.__TS__Class
local __TS__SetDescriptor = ____lualib.__TS__SetDescriptor
local ____exports = {}
____exports.default = __TS__Class()
local Shader = ____exports.default
Shader.name = "Shader"
function Shader.prototype.____constructor(self, gameManager, shader, stopCondition, assets)
    self.isShading = true
    self.elapsedTime = 0
    self.gameManager = gameManager
    self.shader = shader
    self.stopCondition = stopCondition
    self.assets = assets
end
function Shader.prototype.updateShader(self, deltaTime)
    if not self.isShading then
        return
    end
    if self.stopCondition and self:stopCondition() then
        self.isShading = false
        return
    end
    self.elapsedTime = self.elapsedTime + deltaTime
end
__TS__SetDescriptor(
    Shader.prototype,
    "isFinished",
    {get = function(self)
        return not self.isShading
    end},
    true
)
return ____exports

File: Shaders/Shader.ts
import Asset from "Assets/Asset";
import FontWithPosition from "Assets/FontWithPosition";
import GameManager from "GameManager";
import { Shader as LoveShader } from "love.graphics";

export type ShaderAssets = Asset | FontWithPosition;

export default class Shader {
    gameManager: GameManager;
    stopCondition: () => boolean;
    assets: ShaderAssets[];
    isShading: boolean = true;
    elapsedTime = 0;
    shader: LoveShader;
    
    constructor(gameManager: GameManager, shader: LoveShader, stopCondition: () => boolean, assets: ShaderAssets[]) {
        this.gameManager = gameManager;
        this.shader = shader;
        this.stopCondition = stopCondition;
        this.assets = assets;
    }

      updateShader(deltaTime: number): void {
        if (!this.isShading) {
          return;
        }
    
        if (this.stopCondition && this.stopCondition()) {
          this.isShading = false;
          return;
        }

        this.elapsedTime += deltaTime;
      }

      get isFinished(): boolean {
        return !this.isShading;
      }
}